src/coreComponents/linearAlgebra/solvers/PreconditionerBlockJacobi.hpp

```diff
@@ -0,0 +1,177 @@
+#define GEOSX_LINEARALGEBRA_SOLVERS_PRECONDITIONERBLOCKJACOBI_HPP_
+
+#include "linearAlgebra/interfaces/LinearOperator.hpp"
+#include "linearAlgebra/solvers/PreconditionerBase.hpp"
+#include "linearAlgebra/interfaces/BlasLapackLA.hpp"
+
+namespace geosx
+{
+
+/**
+ * @brief Common interface for identity preconditioning operator
+ * @tparam LAI linear algebra interface providing vectors, matrices and solvers
+ */
+template< typename LAI >
+class PreconditionerBlockJacobi : public PreconditionerBase< LAI >
+{
+public:
+
+  /// Alias for base type
+  using Base = PreconditionerBase< LAI >;
+
+  /// Alias for vector type
+  using Vector = typename Base::Vector;
+
+  /// Alias for matrix type
+  using Matrix = typename Base::Matrix;
+
+  /**
+   * @brief Constructor.
+   * @param blockSize the size of block diagonal matrices.
+   */
+  PreconditionerBlockJacobi( localIndex const & blockSize = 0 )
+    : m_blockDiag{}
+  {
+    setBlockSize( blockSize );
+  }
+
+  virtual ~PreconditionerBlockJacobi()
+  {
+    delete m_blockDiag;
+  }
+
+  /**
+   * @brief Set the block size.
+   * @param blockSize the block size.
+   */
+  void setBlockSize( localIndex const & blockSize )
+  {
+    GEOSX_LAI_ASSERT_GT( blockSize, 0 );
+    m_blockSize = blockSize;
+  }
+
+  /**
+   * @brief Compute the preconditioner from a matrix.
+   * @param mat the matrix to precondition.
+   */
+  virtual void compute( Matrix const & mat ) override
+  {
+    GEOSX_LAI_ASSERT( mat.ready() );
+    GEOSX_LAI_ASSERT_GT( m_blockSize, 0 );
+    GEOSX_LAI_ASSERT_EQ( mat.numLocalRows() % m_blockSize, 0 );
+    GEOSX_LAI_ASSERT_EQ( mat.numLocalCols() % m_blockSize, 0 );
+
+    PreconditionerBase< LAI >::compute( mat );
+
+    m_blockDiag = new Matrix();
+    m_blockDiag->createWithLocalSize( mat.numLocalRows(), mat.numLocalCols(), m_blockSize, mat.getComm() );
+    m_blockDiag->open();
+
+    array1d< globalIndex > idxBlk( m_blockSize );
+    array2d< real64 > values( m_blockSize, m_blockSize );
+    array2d< real64 > valuesInv( m_blockSize, m_blockSize );
+    for( globalIndex i = mat.ilower(); i < mat.iupper(); i+=m_blockSize )
+    {
+      values.setValues< serialPolicy >( 0.0 );
+      for( localIndex j = 0; j < m_blockSize; ++j )
+      {
+        globalIndex const iRow = i + LvArray::integerConversion< globalIndex >( j );
+        idxBlk[j] = iRow;
+        localIndex const rowLength = mat.globalRowLength( iRow );
+        array1d< globalIndex > cols( rowLength );
+        array1d< real64 > vals( rowLength );
+        mat.getRowCopy( iRow, cols, vals );
+        for( localIndex k = 0; k < rowLength; ++k )
+        {
+          localIndex const jCol = LvArray::integerConversion< localIndex >( cols[k]-i );
+          if( cols[k] >= i && cols[k] < i+LvArray::integerConversion< globalIndex >( m_blockSize ) )
+          {
+            values( j, jCol ) = vals[k];
+          }
+        }
+      }
+      BlasLapackLA::matrixInverse( values, valuesInv );
```

Comment:
Maybe this is a good long term plan, but I would suggest to keep the current `BlasLapackLA`, being easy to use and quite flexible.

For the specific case at hand, if you provide me an alternative way to invert a dense matrix of size `N x N` (`N` is the block size, that can be larger than 3, being this a general interface), I'll be happy to test it and, eventually, to substitute the `lapack` call.

Reply:
The `tensorOps` stuff will only work if you know the matrix sizes at compile time. It is also only meant for small matrix sizes (small enough that you don't need to thread the operation itself). If neither of these conditions are met you should use the `BlasLapackLA` which **should not** be called from any parallel code. If you have a use case for a `N x N` inverse where you know `N` at compile time and is not too big I'd happily extend the `tensorOps::inverse` functionality.

Reply:
@corbett5 When you say not to call from parallel code, do you mean threaded / GPU code?  I presume calling it on CPU on several parallel mpi ranks is okay?

Reply:
I guess like 99.9% of our code is parallel when run through MPI haha. But yeah I meant threaded / GPU code. You can't make BLAS calls inside a GPU kernel since they're host-only, and you shouldn't make them inside any CPU threaded region since they themselves most likely use threads or even the GPU in the case of IBM's ESSL.

If we start to make big calls to BLAS or LAPACK we should probably link to something like MAGMA which will allow us to make calls where the memory is already on the device.

Reply:
In general, `N` is not known at compile time. So, for now I would use `BlasLapackLA` .

Reply:
Assuming that inside a GPU kernel one: (i) needs to invert matrices (or better factorize + forward/backward substitution), compute QR or SVD decomposition, etc., and (ii) has no compile time knowledge of the size `N`, with `N` small. What are the options available for this scenario?

Reply:
We could look around for a library that does this stuff, I believe `eigen` has routines callable on device. In fact I based the 2x2 and 3x3 eigen solver in `tensorOps` off of their implementation. If we can't find a library to do this it would be easy enough to add runtime sized routines to `tensorOps`. That said GPU performance is much better if you know the loop bounds at compile time so if there's a fixed set of values that need to be supported it might be worth adding another dispatch in.